src/services/volumeMeterService.ts

```diff
@@ -0,0 +1,341 @@
+    let audioSource = this.mediaElementToSourceNodeMap.get(element)
+    if (audioSource) {
+      return audioSource
+    }
+    audioSource = audioContext.createMediaElementSource(element)
+    // `createMediaElementSource` disconnected audio from output,
+    // so connect it back
+    audioSource.connect(audioContext.destination)
+    this.mediaElementToSourceNodeMap.set(element, audioSource)
+    return audioSource
+  }
+
+  private clear (canvas: HTMLCanvasElement) {
+    if (!canvas) {return}
+    const canvasCtx = canvas.getContext('2d')
+    if (!canvasCtx) {return}
+
+    const { width, height } = dimensions(canvas)
+    canvasCtx.clearRect(0, 0, width, height)
+  }
+
+  private drawEmpty (canvas: HTMLCanvasElement) {
+    const canvasCtx = canvas.getContext('2d')
+    if (!canvasCtx) {
+      return
+    }
+
+    const { width, height } = dimensions(canvas)
+    canvasCtx.clearRect(0, 0, width, height)
+    canvasCtx.fillStyle = this.config.backgroundColor
+    canvasCtx.fillRect(0, 0, width, height)
+
+    // Render numbering
+    canvasCtx.fillStyle = this.config.fontColor
+    canvasCtx.font = `${this.config.fontSize}px ${this.config.fontFamily}`
+    const steps = Math.round((
+      this.config.maxDecibels - this.config.minDecibels) / 10)
+    const startStep = Math.round(- this.config.maxDecibels / 10)
+    const heightStep = height / steps
+    for (let i = 1; i < steps; i++) {
+      const label = this.config.levelLabel(-10 * (i + startStep))
+      canvasCtx.fillText(label, 3, i * heightStep)
+    }
+  }
+
+  private draw (canvas: HTMLCanvasElement, leftAudioCtx: AnalyzerContext, rightAudioCtx: AnalyzerContext) {
+    this.drawEmpty(canvas)
+
+    const canvasCtx = canvas.getContext('2d')
+    if (!canvasCtx) {return}
+
+    const { width, height } = dimensions(canvas)
+    const leftBarHeight = this.computeBarHeight(leftAudioCtx, height)
+    const rightBarHeight = this.computeBarHeight(rightAudioCtx, height)
+    const barWidth = this.config.barWidth || (width / 6)
+    const barLeftOffset = this.config.barLeftOffset || (3.4 * barWidth)
+    const barSpacing = this.config.barSpacing
+
+    const gradient = canvasCtx.createLinearGradient(0, height, 0, 0)
+    this.config.barGradient.forEach(({ stop, color }) => {
+      gradient.addColorStop(stop, color)
+    })
+
+    // Render volume bar
+    canvasCtx.fillStyle = gradient
+    canvasCtx.fillRect(barLeftOffset, height - leftBarHeight,
+      barWidth, leftBarHeight)
+    canvasCtx.fillRect(barLeftOffset + barWidth + barSpacing,
+      height - rightBarHeight, barWidth, rightBarHeight)
+  }
+
+  private computeBarHeight (analyzerCtx: AnalyzerContext, maxHeight: number): number {
+    const { analyser, dataArray } = analyzerCtx
+    const MAX_VALUE = 255
+
+    analyser.getByteFrequencyData(dataArray)
+
+    const groupCount = this.config.frequencyBinCount
+    const groupSize = dataArray.length / groupCount
+
+    // Calculate the averages
+    const averages: number[] = []
```

Comment:
I think this can be optimized since this is part of the rendering loop. If I am not mistaken the spec says that the `slice()` call is O(n) (see [here](https://tc39.es/ecma262/multipage/indexed-collections.html#sec-array.prototype.slice) and the `reduce()` call also iterates in O(n). A bit further down we go one more time through the groups to find the `max`.

I think think we can at least remove the second iterations if we do something like (untested, just typing here):

```
let maxAvg = 0
for (let i = 0; i < groupCount; i++) {
  const start = i * groupSize
  const end = (i + 1) * groupSize
  let sumValues = 0
  for (let j=start; j<end; j++) {
    sumValues += dataArray[j]
  }
  maxAvg = Math.max(maxAvg, sumValues / groupSize)
}
 const averagePercent =  maxAvg / MAX_VALUE
```
There is probably a way to make that a bit nicer though I think it makes sense to make sure that we are not wasting anything here since this is a rather tight loop.

Reply:
I absolutely agree this part should be swift, but I'm gonna have to push back here a little bit, since I'm convinced that the time complexity is OK. The current time complexity is `Θ(n)`, you can't really go lower than `n` since you need to iterate over all value of the array.

I'm pretty sure slice is `Θ(n)` meaning `Θ(end-start)`. This article agrees with me https://stackoverflow.com/questions/22614237/javascript-runtime-complexity-of-array-functions.

So maybe if I re-write this in a different way, I can bring down the multiplication factor from cca `Θ(4*n)` to let's say `Θ(2*n)`  (depending on how many steps are internally done by `slice` and `reduce` etc., which I can only guess.

From where I'm standing the best way to ensure it's swift would be by tweaking the `frequencyBinCount` config.

What do you think?